plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMetadataFileOperations.java

```diff
@@ -181,32 +180,11 @@ public void testSelectWithLimit(int numberOfFiles)
             assertUpdate("INSERT INTO test_select_with_limit VALUES ('" + k + "', " + i + ")", 1);
         }
 
-        // org.apache.iceberg.util.ParallelIterable, even if used with a direct executor, schedules 2 * ThreadPools.WORKER_THREAD_POOL_SIZE upfront
-        int icebergManifestPrefetching = 2 * ThreadPools.WORKER_THREAD_POOL_SIZE;
+        int limit = 3;
 
-        assertFileSystemAccesses("SELECT * FROM test_select_with_limit LIMIT 3",
-                ImmutableMultiset.<FileOperation>builder()
-                        .addCopies(new FileOperation(METADATA_JSON, INPUT_FILE_NEW_STREAM), 1)
-                        .addCopies(new FileOperation(SNAPSHOT, INPUT_FILE_GET_LENGTH), 1)
-                        .addCopies(new FileOperation(SNAPSHOT, INPUT_FILE_NEW_STREAM), 1)
-                        .addCopies(new FileOperation(MANIFEST, INPUT_FILE_NEW_STREAM), min(icebergManifestPrefetching, numberOfFiles))
-                        .build());
-
-        assertFileSystemAccesses("EXPLAIN SELECT * FROM test_select_with_limit LIMIT 3",
-                ImmutableMultiset.<FileOperation>builder()
-                        .addCopies(new FileOperation(METADATA_JSON, INPUT_FILE_NEW_STREAM), 1)
-                        .addCopies(new FileOperation(SNAPSHOT, INPUT_FILE_GET_LENGTH), 1)
-                        .addCopies(new FileOperation(SNAPSHOT, INPUT_FILE_NEW_STREAM), 1)
-                        .addCopies(new FileOperation(MANIFEST, INPUT_FILE_NEW_STREAM), numberOfFiles)
-                        .build());
-
-        assertFileSystemAccesses("EXPLAIN ANALYZE SELECT * FROM test_select_with_limit LIMIT 3",
-                ImmutableMultiset.<FileOperation>builder()
-                        .addCopies(new FileOperation(METADATA_JSON, INPUT_FILE_NEW_STREAM), 1)
-                        .addCopies(new FileOperation(SNAPSHOT, INPUT_FILE_GET_LENGTH), 1)
-                        .addCopies(new FileOperation(SNAPSHOT, INPUT_FILE_NEW_STREAM), 1)
-                        .addCopies(new FileOperation(MANIFEST, INPUT_FILE_NEW_STREAM), numberOfFiles + min(icebergManifestPrefetching, numberOfFiles))
-                        .build());
+        assertFileSystemDataFileAccesses("SELECT * FROM test_select_with_limit LIMIT " + limit, min(limit, numberOfFiles));
```

Comment:
@findepi the number of manifest accesses is no longer deterministic, because the splits are enqueued async. As soon as it's identified during a `getNextBatch` call that we have enough splits, the enqueuing will stop, but we can't make any guarantees about the number of manifest accesses.

On the other hand, the number of data file accesses are deterministic and no more data files will be read than necessary. Changed this test to reflect this.

Reply:
thinking about this

- we still want to test manifest accesses;  this is important metric
- there should probably be a kill-switch for this functionality

so maybe the test should just use that kill-switch?

Reply:
I've been thinking about this too, but I think adding a kill switch would complicate the code greatly. The iterative split generation approach is fundamentally different from the background async split loading approach, and a kill switch would require maintaining the code for both approaches in parallel and duplicating the logic in some places (and more painfully, porting fixes to both places). Because of this, I think it's impractical to introduce this flag.

I've made some enhancement to the test though, since even with this new approach we can make some guarantees about the number of manifest accesses. I've changed the test to compare the manifest access count against a lower bound and upper bound, and added some code comments to explain the reasoning behind the bounds. Regardless of the number of manifests in the scan, if there's a limit, the number of actually-accessed manifests should stay within a constant, narrow range.

Reply:
> The iterative split generation approach is fundamentally different from the background async split loading approach

kill switches are useful in case a bug is uncovered. if the new code  is fundamentally different from the old one, it sounds like it's not safe to proceed, is it?

Reply:
I phrased it wrong, I think. They're not fundamentally different, since pretty much all of the split generation and filtering logic has remained unchanged. It's different only in the sense that hiding it behind a feature flag is not trivial like putting some code in an if-else statement. But it's doable. For example, I could make two different classes `IcebergSplitSource` and `IcebergAsyncSplitSource` (with some common logic factored out into a parent base class) and toggle between the two implementations based on the flag - what do you think?

Reply:
Can we make the new implementation produce the same behaviour as before if `max-outstanding-splits` was configurable and set to 1 ?